BraveShared/Analytics/UserReferralProgram.swift

```diff
@@ -189,27 +200,64 @@ public class UserReferralProgram {
             } catch {
                 log.error("Failed to save URP custom header data \(headers) with error: \(error.localizedDescription)")
             }
+            result.fill(headers)
         }
+        
+        return result
+    }
+    
+    /// Returns custom headers synchronously
+    private var customHeaders: [CustomHeaderData]? {
+        guard let customHeadersAsData = Preferences.URP.customHeaderData.value else { return nil }
+        
+        do {
+            return try NSKeyedUnarchiver.unarchiveTopLevelObjectWithData(customHeadersAsData) as? [CustomHeaderData]
+        } catch {
+            log.error("Failed to unwrap custom headers with error: \(error.localizedDescription)")
+        }
+        return nil
     }
     
     /// Checks if a custom header should be added to the request and returns its value and field.
     public class func shouldAddCustomHeader(for request: URLRequest) -> (value: String, field: String)? {
-        guard let customHeadersAsData = Preferences.URP.customHeaderData.value,
-            let customHeaders = (try? NSKeyedUnarchiver.unarchiveTopLevelObjectWithData(customHeadersAsData)) as? [CustomHeaderData],
+        guard let customHeaders = UserReferralProgram()?.customHeaders,
             let hostUrl = request.url?.host else { return nil }
         
         for customHeader in customHeaders {
             // There could be an egde case when we would have two domains withing different domain groups, that would
             // cause to return only the first domain-header it approaches.
-            for domain in customHeader.domainList {
-                if hostUrl.contains(domain) {
-                    if let allFields = request.allHTTPHeaderFields, !allFields.keys.contains(customHeader.headerField) {
-                        return (customHeader.headerValue, customHeader.headerField)
-                    }
+            for domain in customHeader.domainList where hostUrl.contains(domain) {
+                // If `domain` is "cookie only", we exclude it from HTTP Headers, and just use cookie approach
+                let cookieOnly = urpCookieOnlyDomains.contains(domain)
```

Comment:
We are checking all request urls, this means the cookie will be set for any URL that using coinbase assets?  Can we limit this check to urls loaded in top-level frame? Let me know if I'm missing something.

Reply:
Webkit looks for cross domain referencing and does not add for such cases. ie if a some other site beside a domain of coinbase should load a coinbase url then the cookie won't be sent. This is not true for iFrames obviously. Apart from this there is no way to limit cookie sent to document url only.